fix(readability/casting): noexcept *functions

aaronliu0130 opened the pull request:
Fixes #354

this shall be squashed

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection of deprecated C-style pointer casts.
  * Reduced false-positive warnings for function-pointer declarations involving `noexcept`.

* **Tests**
  * Added regression coverage for pointer casts and `noexcept` function-pointer syntax.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aaronliu0130 committed:
fix(readability/casting): noexcept *functions

aaronliu0130 committed:
fix typo

lovewave02 reviewed (changes requested):
Adding `noexcept` to the generic remainder exemption also hides real casts where `noexcept` is the C++ operator. On this head, `bool safe = (bool) noexcept(f());` produces no `readability/casting` diagnostic, while current `develop` reports it and `c++ -std=c++11 -fsyntax-only` accepts the expression.

Please restrict the exemption to a function declaration or function-pointer context and add regression cases for both the intended declaration and the `noexcept(...)` operator expression.

aaronliu0130 committed:
alternative approach to exclude "cast"s that follow closing parenthesis
without space
it seems rare that a cast would directly follow a closing parenthesis
and this excludes function pointers, avoiding the noexcept issue
entirely without accidentally skipping the noexcept operator as well
add unit test for noexcept operator true positive
add TODO on refactoring CheckCStyleCast()

aaronliu0130 commented on cpplint.py:
```diff
@@ -6720,7 +6720,7 @@ def CheckCasts(filename, clean_lines, linenum, error):
         # Check pointer casts for other than string constants
         CheckCStyleCast(
-            filename, clean_lines, linenum, "reinterpret_cast", r"\((\w+\s?\*+\s?)\)", error
+            filename, clean_lines, linenum, "reinterpret_cast", r"(?<!\))\((\w+\s?\*+\s?)\)", error
```
There's a space in your example, so this wouldn't match that; when we're down this niche, isn't it only function pointers that are typically formatted without a space?
How would I exclude only the function pointer shape without needing to import `regex`?

androvonx95 reviewed (commented):
Confirmed this fixes **#354** at `f44a09a`, and `lovewave02`'s `noexcept`-operator cases still warn correctly (`(Type**) noexcept(f())`, `(bool) noexcept(f())`).

**Concern:** the `(?<!\))` lookbehind skips any pointer cast preceded by `)`, not just function pointers. These warn on `develop`, silent here:

```cpp
uintptr_t a = (uintptr_t)(void*)p;
Handle h2 = (Handle)(void*)h;
g(f((Foo)(char*)p));
x = *(Foo)(char*)p;
```

Adding one space (`(uintptr_t) (void*)p`) restores the warning, so it keys off formatting rather than syntax.

**Alternative:** keep the original `\((\w+\s?\*+\s?)\)` pattern, and add this before the remainder check in `CheckCStyleCast()`:

```python
# Parameter lists of function-pointer declarators are not casts, e.g.
#   void (*execute_)(operation_base*) noexcept;
if re.search(r"\((?:[^()]*::)?\s*\*\s*\w*\s*\)\s*$", context):
    return False
```

Tested on current `develop`: fixes **#354**, keeps both `noexcept`-operator true positives, keeps all four lines above warning, covers member function pointers (`void (Cls::*m)(operation_base*) noexcept;`), full suite passes (206 tests).
